libz-rs-sys/src/gz.rs

```diff
@@ -249,12 +281,86 @@ unsafe fn gzopen_help(path: *const c_char, fd: c_int, mode: *const c_char) -> gz
         }
     }
 
-    // FIXME verify the file headers, and initialize the inflate/deflate state
+    // Initialize stream
+    gz_reset(state);
 
     // FIXME change this to core::ptr::from_mut(state).cast::<gzFile_s>() once MSRV >= 1.76
     (state as *mut GzState).cast::<gzFile_s>()
 }
 
+// Reset the internal state of an open gzip stream according to
+// its mode (read or write)
+fn gz_reset(state: &mut GzState) {
+    state.have = 0;                 // no output data available
+    if state.mode == GzMode::GZ_READ {
+        state.eof = false;          // not at end of file
+        state.past = false;         // have not read past end yet
+        state.how = How::Look;      // look for gzip header
+    } else {
+        state.reset = false;        // no deflateReset pending
+    }
+    state.seek = false;             // no seek request pending
+    // Safety: It is valid to pass a null msg pointer to `gz_error`.
+    unsafe { gz_error(state, Z_OK, ptr::null_mut()) };    // clear error status
+    state.pos = 0;                  // no uncompressed data yet
+    // FIXME add once the deflate state is implemented:
+    // state->strm.avail_in = 0;       /* no input data yet */
+}
+
+// Set the error message for a gzip stream, and deallocate any
+// previously set error message.
+//
+// # Safety
+//
+// - `state` must be a properly constructed `GzState`, e.g. as produced by [`gzopen`]
+// - `msg` must be null or a valid C string. If `msg` is non-null, this function will
+//   make a copy if it, and the caller will retain ownership of the original.
+unsafe fn gz_error(state: &mut GzState, err: c_int, msg: *const c_char) {
+    if !state.msg.is_null() {
+        // NOTE: zlib-ng skips the deallocation here if state.err == Z_MEM_ERROR.
+        // Instead, we maintain the invariant that `state.msg`, if non-null, is
+        // allocated using `ALLOCATOR`, and we assume that the heap is coherent
+        // enough that we can (and should) free the old value even if a new alloc
+        // just failed.
+        unsafe { deallocate_cstr(state.msg.cast_mut()) };
+        state.msg = ptr::null_mut();
+    }
```

Comment:
can you elaborate here? My default is to just do what they do. Why should we deviate?